Homework 6

thejohnsondev opened the pull request:
The list of movies is filled from a json file, the data from which is loaded into the MovieRepository class using a coroutine. When you click on an element of the list of movies, a detail fragment with details about the selected film opens. I decided to transfer the data about the selected movie to the detail fragment in the form of an movie.id through the arguments of the fragments. In the detail fragment to display the list of actors, I created a RecyclerView with a horizontal linear layout
![photo_2020-11-25_20-02-20](https://user-images.githubusercontent.com/73844793/104458937-ae333d80-55b4-11eb-9935-d4b4f1bfb3f9.jpg)
![photo_2021-01-13_15-32-26](https://user-images.githubusercontent.com/73844793/104458939-af646a80-55b4-11eb-95bf-422617aa73ee.jpg)

thejohnsondev committed:
HomeWork5(fixed)

I add anotation @DrawableRes to make code clearer, moved hardcoded text to string resources

thejohnsondev committed:
Add coroutines and serialization dependencies, add json files, add 2 model class: Actor.kt and Genre.kt, some code refactor

thejohnsondev committed:
Filling the movieList with data from a JSON file, loading movie posters from the web using glide library, minor changes of layouts

thejohnsondev committed:
Implemented clicking on a list item to open detail fragment with real data using fragment arguments

thejohnsondev committed:
Added a list of actors in a detail fragment with real data

emartynov commented on app/build.gradle:
```diff
@@ -40,8 +41,12 @@ dependencies {
     implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
-    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
-    testImplementation 'junit:junit:4.+'
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.0.1"
```
So single quote or double quote?

thejohnsondev replied:
Исправлю

> Do you want also to clean up project - remove unused images?

Хорошая идея, исправлю.

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/ActorsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+
+
+class ActorsAdapter(
+        private val movie: Movie,
```
Why not pass actors?

thejohnsondev replied:
Вместо Movie передать через конструктор список актеров?

emartynov replied:
Yup

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/FragmentMoviesDetails.kt:
```diff
@@ -4,24 +4,73 @@ import android.os.Bundle
-        val backBtn : TextView = view.findViewById(R.id.back_btn)
-        backBtn.setOnClickListener(){
+        movieId = arguments?.getInt(MOVIE_KEY)
+        currentMovie = MovieRepository.moviesList.find { it.id == movieId }
```
Can you check once that movie is not null and remove !! everywhere after?

thejohnsondev replied:
Проверять currentMovie на null перед использованием где либо, или проверять перед присваиванием его к полю currentMovie как я сделал ниже? 
![Screenshot 2021-01-16 195756](https://user-images.githubusercontent.com/73844793/104819293-345db700-5835-11eb-85db-9a1208ed4026.png)

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/FragmentMoviesList.kt:
```diff
@@ -2,60 +2,74 @@ package com.johnsondev.doboshacademyapp
 class FragmentMoviesList : Fragment() {
 
-    private var spanCount: Int? = null
+    private var spanCount: Int? = null    //I save it because at initialization it is equal to null (studio prompts to do save)
```
Why do you need to save it in the field?

thejohnsondev replied:
Сохранил в поле для того, чтоб передать его в LayoutManager. Он меняет количество элементов в одном рядке в RecyclerView в зависимости от ориентации экрана.

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/FragmentMoviesList.kt:
```diff
@@ -2,60 +2,74 @@ package com.johnsondev.doboshacademyapp
 
-        movieList = view.findViewById<RecyclerView>(R.id.movie_list_rv)
+        movieList = view.findViewById(R.id.movie_list_rv)
         movieList?.adapter = MoviesAdapter(view.context, clickListener)
```
why `?`? Aren't you sure it is there?

thejohnsondev replied:
Студия окрасила RecyclerView в серый цвет, типу что типизация здесь не нужна

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/FragmentMoviesList.kt:
```diff
@@ -2,60 +2,74 @@ package com.johnsondev.doboshacademyapp
+        val fragmentMoviesDetails = FragmentMoviesDetails()
+        fragmentMoviesDetails.arguments = bundleWithMovieId
+
+        movieList?.let { _ ->
```
Why do you need to check movie list view for null?

thejohnsondev replied:
Исправлю

emartynov replied:
Wrong!

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/MovieRepository.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.johnsondev.doboshacademyapp.model.Movie
+
+class MovieRepository {
```
I would move loading movies also in repository. Make Repository also object and use same repository in both fragments

thejohnsondev replied:
Ок, исправлю

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/MoviesAdapter.kt:
```diff
@@ -1,61 +1,70 @@
     private val inflater: LayoutInflater = LayoutInflater.from(context)
 
-    override fun getItemCount(): Int = Movie.movies.size
+    override fun getItemCount(): Int = MovieRepository.moviesList.size
```
Why not to pass movies in the constructor?

thejohnsondev replied:
Ок, исправлю

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/MoviesAdapter.kt:
```diff
@@ -1,61 +1,70 @@
 }
 
-class MovieViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+class MovieViewHolder(view: View, context: Context) : RecyclerView.ViewHolder(view) {
```
You can get context from view

thejohnsondev replied:
Ок, исправлю это

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/MoviesAdapter.kt:
```diff
@@ -1,61 +1,70 @@
+        genre.text = movie.genres.joinToString { it.name }
+        reviews.text = "${movie.reviewCount} $strReviews"
+        rating.progress = movie.rating * 2
+        time.text = "${movie.runningTime} $strMin"
```
Check getString with multiple parameters. So you can make it:
```
context.getString(R.string.running_time, numberOfMinutes, minutes)
```

But actually you don't need even do such. Just have string with placeholder in strings:
```
<string id="running_time">%d mins</string>
```

emartynov commented on app/src/main/res/values/dimens.xml:
```diff
@@ -20,5 +20,8 @@
     <dimen name="dp_2">2dp</dimen>
     <dimen name="sp_8">8sp</dimen>
     <dimen name="dp_14">14dp</dimen>
+    <dimen name="image_height">248dp</dimen>
```
Why such hardcoded values?

thejohnsondev replied:
Без этого dimension изображение либо сжимается, либо растягивается, и получается не красиво

emartynov replied:
start_to_start=parent
end_to_end=parent
contraintRatio="3:2"

emartynov replied:
centerInside

emartynov replied:
Не получилось?

thejohnsondev replied:
В detail fagment пофиксил, просто этот дименшн использовался в другом лейауте. Теперь пофиксил везде)

emartynov reviewed (commented):
Some places to improve

emartynov commented:
Do you want also to clean up project - remove unused images?

emartynov commented on app/src/main/res/values/strings.xml:
```diff
@@ -16,4 +16,9 @@
     <string name="back">Back</string>
     <string name="movies_list">Movies list</string>
     <string name="_137_min">137 MIN</string>
```
<string name="duration_format">%d MIN</string>

val durationValue = movie.duration // Int

val durationString = getString(R.string.duration_format, duratioValue)

thejohnsondev committed:
Homework fix

emartynov commented on app/src/main/java/com/johnsondev/doboshacademyapp/ActorsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+    private val actorImg: ImageView = view.findViewById(R.id.actor_img)
+    private val actorName: TextView = view.findViewById(R.id.tv_actor_name)
+
+    fun bind(actorsList: List<Actor>, position: Int) {
```
А чего сразу актера не передавать?

thejohnsondev replied:
Пофиксил

emartynov reviewed (approved):
Оке, один комент оставил еще

thejohnsondev committed:
Homework fix #2

The height of the imageViews is calculated using the constraintDimensionRatio, a specific actor is passed in the bind method

emartynov reviewed (approved):
LGTM!
